pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/value/FirstValueWindowFunction.java

```diff
@@ -78,12 +86,73 @@ private List<Object> processRowsWindow(List<Object[]> rows) {
       lowerBound++;
       upperBound = Math.min(upperBound + 1, numRows - 1);
     }
+
+    return result;
+  }
+
+  private List<Object> processRowsWindowIgnoreNulls(List<Object[]> rows) {
```

Comment:
Should we short circuit the unbounded case?

Reply:
I don't think there's much benefit to that here? In the unbounded case for `FIRST_VALUE` / `LAST_VALUE` with `IGNORE NULLS`, we'll find the first / last non-null value in the first window (which will encompass all rows) at the beginning and then in each iteration there's only some simple boolean checks which will all be `false` in every iteration of the loop and we'll simply keep adding the same value to the result list.

Reply:
We can replace the for loop to nCopy, and also short-circuit all if checks. The total cost of the query is just finding the first/last non-null value, so the save could potentially be relatively significant.

From high level, for `UNBOUNDED PROCEEDING, CURRENT ROW, UNBOUNDED FOLLOWING`, the behavior of `ROWS` and `RANGES` should be the same. So we could potentially split the handling into 3 cases:
- Without bounded proceeding/following
- ROWS with bounded proceeding/following
- RANGE with bounded proceeding/following - not supported

We can discuss and address this in a separate PR

Reply:
> From high level, for UNBOUNDED PROCEEDING, CURRENT ROW, UNBOUNDED FOLLOWING, the behavior of ROWS and RANGES should be the same

The behavior will only be the same for `UNBOUNDED PRECEDING TO UNBOUNDED FOLLOWING` right? With `CURRENT ROW` as either lower or upper bound or both, the behavior and logic are both significantly different due to the need to consider peer groups (rows before and after with the same order key) along with current row with `RANGE` type window frames. 

> We can replace the for loop to nCopy, and also short-circuit all if checks. The total cost of the query is just finding the first/last non-null value, so the save could potentially be relatively significant.

This makes sense for true unbounded case though (`UNBOUNDED PRECEDING TO UNBOUNDED FOLLOWING`, `CURRENT ROW` is not really unbounded), I've raised a small follow-up PR - https://github.com/apache/pinot/pull/14324.

Reply:
I see. Good point on the peer group difference